### PR TITLE
add a fallback for kube-scheduler  when events.k8s.io is disabled

### LINKED ISSUE #83203: kube-scheduler should fallback to core/v1 Events if events.k8s.io/v1beta1 is unavailable
In v1.16.0 https://github.com/kubernetes/kubernetes/pull/78447 migrated kube-scheduler to emit events against the v1beta1 API, but didn't implement a fallback to core/v1 in case `events.k8s.io/v1beta1` is not available. we should fix this.

The fix would targed v1.16.2

/assign @yastij 
cc @wojtek-t @liggitt 

/sig scheduling
/priority important-soon

### PR DESCRIPTION
**Pull request by yastij**
Signed-off-by: Yassine TIJANI <ytijani@vmware.com>

**What type of PR is this?**

/kind bug

**What this PR does / why we need it**: This adds a fallback to kube-scheduler when `events.k8s.io` v1beta1 is disabled. scheduler will fallback to core/v1

**Which issue(s) this PR fixes**: Fixes https://github.com/kubernetes/kubernetes/issues/83203

**Special notes for your reviewer**:

this is a WIP. There are two approaches:

- embedding two recorders in the scheduler
- reworking the `events/` package to implement a fallback mechanism

this PR implements the first option. The second one would require reworking extensively the new implementation, but open for suggestions.

/assign @liggitt @wojtek-t 

**Does this PR introduce a user-facing change?**:

```release-note
kube-scheduler now fallbacks to emitting events using core/v1 Events when events.k8s.io/v1beta1 is disabled.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: cmd/kube-scheduler/app/options/options.go [hunk 1]
```diff
@@ -237,15 +238,19 @@ func (o *Options) Config() (*schedulerappconfig.Config, error) {
 	}
 
 	// Prepare event clients.
-	eventBroadcaster := events.NewBroadcaster(&events.EventSinkImpl{Interface: eventClient.EventsV1beta1().Events("")})
-	recorder := eventBroadcaster.NewRecorder(scheme.Scheme, c.ComponentConfig.SchedulerName)
-	leaderElectionBroadcaster := record.NewBroadcaster()
-	leaderElectionRecorder := leaderElectionBroadcaster.NewRecorder(scheme.Scheme, corev1.EventSource{Component: c.ComponentConfig.SchedulerName})
+	if _, err := client.Discovery().ServerResourcesForGroupVersion(eventsv1beta1.SchemeGroupVersion.String()); err == nil {
```
Comment (liggitt) [inline cmd/kube-scheduler/app/options/options.go]: were we making other API calls to the server in `Config()` previously?

File: cmd/kube-scheduler/app/options/options.go [hunk 2]
```diff
@@ -237,15 +238,19 @@ func (o *Options) Config() (*schedulerappconfig.Config, error) {
...
+		c.Broadcaster = events.NewBroadcaster(&events.EventSinkImpl{Interface: eventClient.EventsV1beta1().Events("")})
+		c.Recorder = c.Broadcaster.NewRecorder(scheme.Scheme, c.ComponentConfig.SchedulerName)
+		c.EventClient = eventClient.EventsV1beta1()
```
Comment (liggitt) [inline cmd/kube-scheduler/app/options/options.go]: do all call sites tolerate these three fields being nil?

Comment (yastij) [reply cmd/kube-scheduler/app/options/options.go]: not yet, but idea is to do so

File: cmd/kube-scheduler/app/options/options.go (hunk 1 above)
Comment (yastij) [reply cmd/kube-scheduler/app/options/options.go]: no we were not IIRC. If there's an error the v1beta1 related fields would be nil -> fallback. Are you concerned about adding an API call for building sheduler's config ?

Comment (wojtek-t) [reply cmd/kube-scheduler/app/options/options.go]: @liggitt - can you explain why do you think it's a problem?

Comment (liggitt) [reply cmd/kube-scheduler/app/options/options.go]: It removes our ability to construct Config without a live server present. It's normal to make config, then start the process, and expect API calls to not start happening or being required until Run is called

Comment (wojtek-t) [reply cmd/kube-scheduler/app/options/options.go]: @yastij - I see that you already moved that check to server.go - please remove it from here now.

File: pkg/scheduler/factory.go [hunk 3]
```diff
@@ -93,6 +94,9 @@ type Config struct {
 	// Recorder is the EventRecorder to use
 	Recorder events.EventRecorder
```
Comment (liggitt) [inline pkg/scheduler/factory.go]: this is already an interface... why would we not implement this interface using the v1 Events API?

File: pkg/scheduler/scheduler.go [hunk 4]
```diff
@@ -103,6 +104,10 @@ type Scheduler struct {
 	// Recorder is the EventRecorder to use
 	Recorder events.EventRecorder
```
Comment (liggitt) [inline pkg/scheduler/scheduler.go]: same question... a v1 Events implementation of the EventRecorder interface would make more sense

File: cmd/kube-scheduler/app/server.go [hunk 5]
```diff
@@ -168,11 +171,18 @@ func Run(ctx context.Context, cc schedulerserverconfig.CompletedConfig, registry
 		}
 	}
 
+	// Prepare event clients.
+	if _, err := cc.Client.Discovery().ServerResourcesForGroupVersion(eventsv1beta1.SchemeGroupVersion.String()); err == nil {
+		cc.Broadcaster = events.NewBroadcaster(&events.EventSinkImpl{Interface: cc.EventClient.Events("")})
+		cc.Recorder = cc.Broadcaster.NewRecorder(scheme.Scheme, cc.ComponentConfig.SchedulerName)
+	}
```
Comment (liggitt) [inline cmd/kube-scheduler/app/server.go]: I expected:
```go
if _, err := cc.Client.Discovery().ServerResourcesForGroupVersion(eventsv1beta1.SchemeGroupVersion.String()); err == nil {
  cc.Broadcaster = events.NewBroadcaster(&events.EventSinkImpl{Interface: cc.EventClient.Events("")})
} else {
  cc.Broadcaster = ... v1 events-based implementation of events.EventBroadcaster
}
cc.Recorder = cc.Broadcaster.NewRecorder(scheme.Scheme, cc.ComponentConfig.SchedulerName)
```

File: pkg/scheduler/scheduler.go [hunk 6]
```diff
@@ -422,7 +429,11 @@ func (sched *Scheduler) Run(ctx context.Context) {
 func (sched *Scheduler) recordSchedulingFailure(podInfo *framework.PodInfo, err error, reason string, message string) {
 	sched.Error(podInfo, err)
 	pod := podInfo.Pod
-	sched.Recorder.Eventf(pod, nil, v1.EventTypeWarning, "FailedScheduling", "Scheduling", message)
```
Comment (liggitt) [inline pkg/scheduler/scheduler.go]: I expected the fallback logic here: https://github.com/kubernetes/kubernetes/pull/83692/files#r339204676

not in if/else blocks at event record time

File: cmd/kube-scheduler/app/server.go (hunk 5 above)
Comment (liggitt) [reply cmd/kube-scheduler/app/server.go]: with something like this:
```
type v1RecorderWrapper struct {
	recorder record.EventRecorder
}

func (w v1RecorderWrapper) Eventf(regarding runtime.Object, related runtime.Object, eventtype, reason, action, note string, args ...interface{}) {
	w.recorder.Eventf(regarding, eventtype, reason, note, args...)
}
```

Comment (wojtek-t) [reply cmd/kube-scheduler/app/server.go]: +1 to what Jordan wrote
Especially given that Recorder interface is so simple that it has only 1 method:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/events/interfaces.go#L25

Comment (wojtek-t) [reply cmd/kube-scheduler/app/server.go]: I opened https://github.com/kubernetes/kubernetes/pull/84442 - @yastij feel free to patch it into your PR

**Comment by wojtek-t** [comment]
/approve

I will let @liggitt to lgtm it if he's fine with it (this now looks how I would expect).

File: pkg/scheduler/factory.go [hunk 7]
```diff
@@ -93,6 +94,9 @@ type Config struct {
...
 
+	// CoreRecorder is a fallback recorder in case events.k8s.io is disabled
+	CoreRecorder record.EventRecorder
```
Comment (liggitt) [inline pkg/scheduler/factory.go]: this is unused now, right? can we drop this change?

File: pkg/scheduler/scheduler.go [hunk 8]
```diff
@@ -474,7 +474,6 @@ func (sched *Scheduler) preempt(ctx context.Context, state *framework.CycleState
 				waitingPod.Reject("preempted")
 			}
 			sched.Recorder.Eventf(victim, preemptor, v1.EventTypeNormal, "Preempted", "Preempting", "Preempted by %v/%v on node %v", preemptor.Namespace, preemptor.Name, nodeName)
-
```
Comment (liggitt) [inline pkg/scheduler/scheduler.go]: nit: revert no-op change

**Comment by liggitt** [comment]
two nits, lgtm otherwise

**Comment by yastij** [comment]
@liggitt @wojtek-t - updated

File: staging/src/k8s.io/client-go/tools/record/event.go [hunk 9]
```diff
@@ -132,6 +132,25 @@ type EventBroadcaster interface {
...
+// interface.
+func NewEventRecorderAdapter(recorder EventRecorder) *EventRecorderAdapter {
+	return &EventRecorderAdapter{
+		recorder: recorder,
+	}
+}
+
+// Eventf is a wrapper around v1 Eventf
+func (a *EventRecorderAdapter) Eventf(regarding, _ runtime.Object, eventtype, reason, action, note string, args ...interface{}) {
+	a.recorder.Eventf(regarding, eventtype, reason, note, args)
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/tools/record/event.go]: shouldn't this be:
```suggestion
	a.recorder.Eventf(regarding, eventtype, reason, note, args...)
```

**Comment by wojtek-t** [comment]
/lgtm

Thanks!

**Pull request merged**